assembly/src/todo/mod.rs

```diff
@@ -0,0 +1,209 @@
+            };
+
+            context.locals.push(proc);
+        }
+
+        Ok(context)
+    }
+}
+
+pub struct Assembler {
+    module_provider: Box<dyn ModuleProvider>,
+}
+
+impl Assembler {
+    pub fn new() -> Self {
+        Self {
+            module_provider: Box::new(()),
+        }
+    }
+
+    pub fn with_module_provider<P>(mut self, provider: P) -> Self
+    where
+        P: ModuleProvider + 'static,
+    {
+        self.module_provider = Box::new(provider);
+        self
+    }
+
+    pub fn compile<S>(&self, source: S) -> Result<Program, AssemblerError>
+    where
+        S: AsRef<str>,
+    {
+        let source = source.as_ref();
+        let ProgramAst { local_procs, body } = parsers::parse_program(source)?;
+
+        AssemblerContext::new(self, local_procs.iter())
+            .and_then(|mut context| self.compile_body(&mut context, body.iter()))
+            .map(Program::new)
+    }
+
+    fn compile_body<A, N>(
+        &self,
+        context: &mut AssemblerContext,
+        ast: A,
+    ) -> Result<CodeBlock, AssemblerError>
+    where
+        A: Iterator<Item = N>,
+        N: Borrow<Node>,
+    {
+        let (size, hint) = ast.size_hint();
+        let size = hint.unwrap_or(size);
+
+        let mut blocks: Vec<CodeBlock> = Vec::with_capacity(size);
+        let mut span: Vec<Operation> = Vec::with_capacity(size);
+
+        for node in ast {
+            match node.borrow() {
+                Node::Instruction(instruction) => {
+                    context.compile_instruction(&mut blocks, &mut span, instruction)?;
+                }
+
+                Node::IfElse(t, f) => {
+                    span_to_block(&mut blocks, &mut span);
+
+                    let t = self.compile_body(context, t.iter())?;
+                    let f = self.compile_body(context, f.iter())?;
+                    let block = CodeBlock::new_split(t, f);
+
+                    blocks.push(block);
+                }
+
+                Node::Repeat(n, nodes) => {
+                    span_to_block(&mut blocks, &mut span);
+
+                    let block = self.compile_body(context, nodes.iter())?;
+
+                    for _ in 0..*n {
+                        blocks.push(block.clone());
+                    }
+                }
+
+                Node::While(nodes) => {
+                    span_to_block(&mut blocks, &mut span);
+
+                    let block = self.compile_body(context, nodes.iter())?;
+                    let block = CodeBlock::new_loop(block);
+
+                    blocks.push(block.clone());
+                }
+            }
+        }
+
+        if !span.is_empty() {
+            blocks.push(CodeBlock::new_span(span));
+        }
+
+        Ok(parsers::combine_blocks(blocks))
+    }
+}
+
+fn span_to_block(blocks: &mut Vec<CodeBlock>, span: &mut Vec<Operation>) {
+    let span: Vec<_> = span.drain(..).collect();
+    if !span.is_empty() {
+        blocks.push(CodeBlock::new_span(span));
+    }
```

Comment:
minor nit: wouldn't it be better to do it like this?
```Rust
if !span.is_empty() {
    let span: Vec<_> = span.drain(..).collect();
    blocks.push(CodeBlock::new_span(span));
}
```

Reply:
Fixed in span builder suggestion